airflow/executors/python_executor.py

```diff
@@ -0,0 +1,173 @@
+                self.change_state(key, State.FAILED)
+                self.log.exception("Failed to execute task: %s.", str(e))
+
+        self.tasks_to_run = []
+
+    def queue_task_instance(
+        self,
+        task_instance,
+        mark_success=False,
+        pickle_id=None,
+        ignore_all_deps=False,
+        ignore_depends_on_past=False,
+        ignore_task_deps=False,
+        ignore_ti_state=False,
+        pool=None,
+        cfg_path=None,
+    ):
+
+        self.queue_command(
+            task_instance,
+            (
+                task_instance,
+                dict(
+                    mark_success=mark_success,
+                    # ignore_all_deps =ignore_all_deps,
+                    # ignore_depends_on_past=ignore_depends_on_past,
+                    # ignore_task_deps=ignore_task_deps,
+                    # ignore_ti_state=ignore_ti_state,
+                    pool=pool,
+                    # pickle_id=pickle_id,
+                ),
+            ),
+            priority=task_instance.task.priority_weight_total,
+            queue=task_instance.task.queue,
+        )
+
+    def end(self):
+        # we are not async executor
+        # so we can just return from the execution here
+        for key, command in self.tasks_to_run:
+            ti = command[0]
+            logger.info("Setting %s to %s", key, State.UPSTREAM_FAILED)
+            ti.set_state(State.UPSTREAM_FAILED)
+            self.change_state(key, State.UPSTREAM_FAILED)
+        return
+        # self.heartbeat()
+
+    def terminate(self):
+        self._is_terminated.set()
+
+    # overriding default implementation with better logging messages
+    def change_state(self, key, state):
+        logger.debug("popping: {}".format(key))
+        self.running.pop(key)
+        self.event_buffer[key] = state
+
+    def queue_command(self, task_instance, command, priority=1, queue=None):
+        key = task_instance.key
+        if key not in self.queued_tasks and key not in self.running:
+            self.log.debug("Adding to queue: %s", command)
+            self.queued_tasks[key] = (command, priority, queue, task_instance)
+        else:
+            self.log.info("could not queue task {}".format(key))
+
+    # EOF override
+
+    @provide_session
+    def _run_task_instance(self, ti, mark_success, pool, session=None):
+        # set proper state and try number to keep logger in sync
+        ti.state = State.RUNNING
+        ti._try_number += 1
+        # let save state
+        session.merge(ti)
+        session.commit()
+        # backward compatible with airflow loggers
+        from airflow.utils.log import logging_mixin
+
+        logging_mixin.set_context(logging.root, ti)
+
+        try:
+            ti._run_raw_task(mark_success=mark_success, job_id=ti.job_id, pool=pool)
+        finally:
+            for handler in logging.root.handlers:
+                if handler.name == "task":
+                    handler.close()
```

Comment:
What is the reason behind setting context?

Reply:
Initially we used our own logging system instead of what airflow provides, now we can just remove this lines